### PR TITLE
Allow users to change their own usernames

### PR DESCRIPTION
**Pull request by charmander**
Implements #737.

### CODE REVIEW

File: libweasyl/libweasyl/models/tables.py [hunk 1]
```diff
@@ -908,6 +908,33 @@ def default_fkey(*args, **kwargs):
...
+    CheckConstraint("active = (deactivated_at IS NULL) AND active = (deactivated_by IS NULL)", name='username_history_active_check'),
+)
+
+# enforces one active redirect per user
+Index('ind_username_history_userid', username_history.c.userid, postgresql_where=username_history.c.active, unique=True)
+
+# enforces that active redirects have unique usernames within this table, although they also need to be unique in all of login, logincreate, useralias, and username_history together
+Index('ind_username_history_login_name', username_history.c.login_name, postgresql_where=username_history.c.active, unique=True)
+
+# lookup for a user’s most recent change
```
Comment (kfkitsune) [inline libweasyl/libweasyl/models/tables.py]:
```suggestion
# lookup for a user's most recent change
```
Alembic doesn't like the smart quote here, and it's blocking migration via ``upgrade-db`` when provisioning via a ``SyntaxError`` due to an encoding not being declared. A small portion of the traceback is below.

```
    default: make[1]: Entering directory '/vagrant/libweasyl'
    default: ve/bin/alembic upgrade head
    default: Traceback (most recent call last):
    default:   File "ve/bin/alembic", line 8, in <module>
...
    default: from libweasyl.models.tables import metadata as target_metadata
    default:   File "
    default: /vagrant/libweasyl/libweasyl/models/tables.py
    default: ", line
    default: 934
    default: SyntaxError
    default: :
    default: Non-ASCII character '\xe2' in file /vagrant/libweasyl/libweasyl/models/tables.py on line 934, but no encoding declared; see http://python.org/dev/peps/pep-02
    default: Makefile:44: recipe for target 'upgrade-db' failed
    default: make[1]: Leaving directory '/vagrant/libweasyl'
    default: make[1]: *** [upgrade-db] Error 1
    default: Makefile:67: recipe for target 'upgrade-db' failed
    default: make: *** [upgrade-db] Error 2
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
kyra@kitty ~/g/weasyl> cd libweasyl/libweasyl/models/
kyra@kitty ~/g/w/l/l/models> python
Python 2.7.13 (default, Sep 26 2018, 18:42:22)
[GCC 6.3.0 20170516] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> with open("tables.py") as fp:
...     for i, line in enumerate(fp):
...             if "\xe2" in line:
...                     print i, repr(line)
...
933 '# lookup for a user\xe2\x80\x99s most recent change\n'
```

File: weasyl/profile.py [hunk 2]
```diff
@@ -788,28 +781,13 @@ def do_manage(my_userid, userid, username=None, full_name=None, catchphrase=None
...
-
-        with d.engine.begin() as db:
-            db.execute(
-                "UPDATE login SET login_name = %(sysname)s WHERE userid = %(user)s",
-                sysname=sysname, user=userid)
-            db.execute(
-                "UPDATE profile SET username = %(username)s WHERE userid = %(user)s",
-                username=username, user=userid)
-        d._get_display_name.invalidate(userid)
+        login.change_username(
```
Comment (kfkitsune) [inline weasyl/profile.py]: The way this is coded has an interesting effect. If an admin "updates" the username to the same value (e.g., checks the box, doesn't change the contents), this triggers the cooldown for the username change as if the user made the change themselves, and also states that they can release their currently active username.

I can't tell if this is intended to operate in this fashion, but it's certainly an edge case. (And morbid curiosity did lead me to hit ``Release Old Username``; nothing breaks, thankfully.)


Wzl-usernamechangeglitch

Comment (charmander) [reply weasyl/profile.py]: Oh, good point. I originally had code to allow changes that didn’t affect the resulting `login_name` without adding history entries and took it out for being unnecessary complexity, but requiring the username to actually be different at all is very important (for both normal users and admins).

File: weasyl/login.py [hunk 3]
```diff
@@ -328,10 +363,101 @@ def username_exists(login_name):
...
+
+def change_username(acting_user, target_user, bypass_limit, new_username):
+    new_username = clean_display_name(new_username)
+    new_sysname = d.get_sysname(new_username)
+
+    old_username = d.get_display_name(target_user)
+    old_sysname = d.get_sysname(old_username)
+
+    if new_username == old_username:
+        return
```
[the remark this answers is not in the document]
Comment (kfkitsune) [reply weasyl/login.py]: Additional observation post comment: https://gitter.im/Weasyl/weasyl?at=5e4decde8e726c7dc5b61a4a

>  I don't know if the admin interface should use aliases as well, but /modcontrol/manageuser, /modcontrol/reports also don't forward resolve. /admincontrol/finduser may or may not, the branch isn't synced to master.

**Comment by kfkitsune** [comment]
Additional observation after the review was posted: https://gitter.im/Weasyl/weasyl?at=5e4decde8e726c7dc5b61a4a

>  I don't know if the admin interface should use aliases as well, but /modcontrol/manageuser, /modcontrol/reports also don't forward resolve. /admincontrol/finduser may or may not, the branch isn't synced to master.

**Comment by charmander** [comment]
@kfkitsune Thanks for catching the inconsistency! It’s fixed with the latest changes. As for allowing aliases everywhere in the mod tools… I’ll just leave it alone until it comes up. :P

File: weasyl/moderation.py [hunk 4]
```diff
@@ -470,74 +470,57 @@ def setusermode(userid, form):
...
-          FROM journal
-               JOIN profile USING (userid)
-               JOIN login USING (userid)
-         WHERE login_name = %(sysname)s
-    """, sysname=d.get_sysname(form.name)).fetchall()
+def journalsbyuser(targetid):
+    query = d.engine.execute("""
+        SELECT journalid, title, settings, unixtime, rating
+        FROM journal
+        WHERE jo.userid = %(user)s
```
Comment (kfkitsune) [inline weasyl/moderation.py]:
```suggestion
        WHERE userid = %(user)s
```

And here it looks like the ``jo`` got left behind.

https://paste.weasyldev.com/show/bSlGk3PwosxSZivsmnUC/
```
2020-02-21T01:06:07+0000 [stderr#error] ProgrammingError: (psycopg2cffi._impl.exceptions.ProgrammingError) missing FROM-clause entry for table "jo"
2020-02-21T01:06:07+0000 [stderr#error] LINE 4:         WHERE jo.userid = 36064
2020-02-21T01:06:07+0000 [stderr#error]                       ^
2020-02-21T01:06:07+0000 [stderr#error] [SQL:
2020-02-21T01:06:07+0000 [stderr#error]         SELECT journalid, title, settings, unixtime, rating
2020-02-21T01:06:07+0000 [stderr#error]         FROM journal
2020-02-21T01:06:07+0000 [stderr#error]         WHERE jo.userid = %(user)s
2020-02-21T01:06:07+0000 [stderr#error]     ]
```

File: weasyl/moderation.py [hunk 5]
```diff
@@ -470,74 +470,57 @@ def setusermode(userid, form):
...
-            ch.char_name, ch.age, ch.gender, ch.height, ch.weight, ch.species,
-            ch.content, ch.rating, ch.settings, ch.page_views, pr.config
+            ch.charid, pr.username, ch.unixtime, ch.char_name, ch.rating,
+            ch.settings
         FROM character ch
-        INNER JOIN profile pr ON ch.userid = pr.userid
-        INNER JOIN login ON ch.userid = login.userid
-        WHERE login.login_name = %(sysname)s
-    """, sysname=d.get_sysname(form.name))
+            INNER JOIN profile USING (userid)
```
Comment (kfkitsune) [inline weasyl/moderation.py]:
```suggestion
            INNER JOIN profile pr USING (userid)
```

Looks like the ``pr`` got dropped. Trying to use ``/modcontrol/contentbyuser`` will fail as currently coded.

https://paste.weasyldev.com/show/4u1g4a2Vde6juT1CqYV0/
```
2020-02-21T01:00:32+0000 [stderr#error] ProgrammingError: (psycopg2cffi._impl.exceptions.ProgrammingError) missing FROM-clause entry for table "pr"
2020-02-21T01:00:32+0000 [stderr#error] LINE 3:             ch.charid, pr.username, ch.unixtime, ch.char_nam...
2020-02-21T01:00:32+0000 [stderr#error]                                ^
2020-02-21T01:00:32+0000 [stderr#error] [SQL:
2020-02-21T01:00:32+0000 [stderr#error]         SELECT
2020-02-21T01:00:32+0000 [stderr#error]             ch.charid, pr.username, ch.unixtime, ch.char_name, ch.rating,
2020-02-21T01:00:32+0000 [stderr#error]             ch.settings
2020-02-21T01:00:32+0000 [stderr#error]         FROM character ch
2020-02-21T01:00:32+0000 [stderr#error]             INNER JOIN profile USING (userid)
2020-02-21T01:00:32+0000 [stderr#error]         WHERE ch.userid = %(user)s
2020-02-21T01:00:32+0000 [stderr#error]     ]
```

**Pull request merged**